apigee/registry
# Conformance command to compute Lint Results based on API Style Guides

## Summary

This pull request aims to add the `conformance` subcommand under the `compute` command in the registry CLI. This command, when passed the path of an API Spec, will compute and attach Lint objects as artifacts to the spec according to the API Style Guides defined on the project that contains the spec.

To go into detail, consider the following scenario: there are three API style guides present on the project named `example` that govern OpenAPI, AsyncAPI, and GraphQL specs. Then, the user calls 
```bash
registry compute conformance projects/example/locations/global/apis/-/versions/-/specs/-
```
After this, the three style guides under the project are looked up. Each linter is configured to use the rules defined for it across all three style guides. For each spec, the mime type is looked up. If the mime type matches the mime type a linter supports (OpenAPI, AsyncAPI, or GraphQL), the linter is called on the spec, and the resulting Lint object is stored as an artifact on the spec.

## Note
Currently, the Lint objects being attached are blank objects. In a subsequent pull request, the logic of the Spectral and API Linter plugins will be implemented, which will do the actual linting and populate the lint objects.

## Test

- Have registry running on `localhost:8080`.
- Ensure that https://github.com/APIs-guru/openapi-directory is extracted to your Desktop.
- Ensure that the following style guide is located under `~/styleguide.yaml`:
```yaml
name: openapistyleguide
mime_types:
  - application/x.openapi+gzip;version=2
guidelines:
  - name: refproperties
    display_name: Govern Ref Properties
    description: This guideline governs properties for ref fields on specs.
    rules:
      - name: norefsiblings
        description: An object exposing a $ref property cannot be further extended with additional properties.
        linter: spectral
        linter_rulename: no-$ref-siblings
        severity: 1
    status: 1
linters:
  - name: spectral
    uri: https://github.com/stoplightio/spectral
```

Now, run the following series of bash commands:
```bash
apg registry delete-project --name projects/example

apg registry create-project --project_id example --insecure --address localhost:8080

registry upload bulk openapi --project_id example ~/Desktop/openapi-directory/APIs

registry upload styleguide ~/styleguide.yaml --project_id=example

registry compute conformance projects/example/locations/global/apis/zappiti.com/versions/4.15.174/specs/swagger.yaml

registry list projects/example/locations/global/apis/zappiti.com/versions/4.15.174/specs/-/artifacts/-
```

**Expected output**:
```bash
projects/example/locations/global/apis/zappiti.com/versions/4.15.174/specs/swagger.yaml/artifacts/lint-spectral
```
